File: libs/shared/models/typescript/deportes.ts
/**
 * Shared TypeScript interfaces for the Deportes (Sports) module.
 * Mirrors backend Pydantic schemas in apps/backend/src/models/schemas.py.
 */

// ---------------------------------------------------------------------------
// Sport
// ---------------------------------------------------------------------------

export interface SportBase {
  name: string;
  icon?: string;
}

export interface SportCreate extends SportBase {}

export interface SportResponse extends SportBase {
  id: number;
  is_custom: boolean;
  created_at: string;
}

// ---------------------------------------------------------------------------
// Sport Activity
// ---------------------------------------------------------------------------

export interface SportActivityBase {
  sport_id: number;
  date: string;
  duration_seconds: number;
  distance_km?: number;
  calories?: number;
  heart_rate_avg?: number;
  extra_data?: Record<string, unknown>;
  notes?: string;
}

export interface SportActivityCreate extends SportActivityBase {}

export interface SportActivityUpdate {
  date?: string;
  duration_seconds?: number;
  distance_km?: number;
  calories?: number;
  heart_rate_avg?: number;
  extra_data?: Record<string, unknown>;
  notes?: string;
}

export interface SportActivityResponse extends SportActivityBase {
  id: number;
  user_id: number;
  created_at: string;
}

// ---------------------------------------------------------------------------
// Sport Stats & Personal Records
// ---------------------------------------------------------------------------

export interface SportStatsResponse {
  sport_id: number;
  sport_name: string;
  total_activities: number;
  total_time_seconds: number;
  total_distance_km?: number;
  average_duration_seconds?: number;
}

export interface PersonalRecordResponse {
  sport_id: number;
  best_time_seconds?: number;
  best_distance_km?: number;
  best_pace_seconds_per_km?: number;
  best_time_date?: string;
  best_distance_date?: string;
}

// ---------------------------------------------------------------------------
// Training Plan
// ---------------------------------------------------------------------------

export interface TrainingPlanWeekUpdate {
  week_number: number;
  target_sessions: number;
  completed_sessions?: number;
  notes?: string;
}

export interface TrainingPlanBase {
  name: string;
  description?: string;
  start_date: string;
  end_date: string;
}

export interface TrainingPlanCreate extends TrainingPlanBase {
  weeks?: TrainingPlanWeekUpdate[];
}

export interface TrainingPlanUpdate {
  name?: string;
  description?: string;
  start_date?: string;
  end_date?: string;
  weeks?: TrainingPlanWeekUpdate[];
}

export interface TrainingPlanWeekResponse {
  id: number;
  week_number: number;
  target_sessions: number;
  completed_sessions: number;
  notes?: string;
}

export interface TrainingPlanResponse extends TrainingPlanBase {
  id: number;
  created_at: string;
  updated_at: string;
}

export interface TrainingPlanDetailResponse extends TrainingPlanResponse {
  weeks: TrainingPlanWeekResponse[];
}

// ---------------------------------------------------------------------------
// Race Event
// ---------------------------------------------------------------------------

export interface RaceEventBase {
  name: string;
  sport_id: number;
  event_date: string;
  distance_km?: number;
  location?: string;
  target_time_seconds?: number;
  notes?: string;
}

export interface RaceEventCreate extends RaceEventBase {}

export interface RaceEventUpdate {
  name?: string;
  sport_id?: number;
  event_date?: string;
  distance_km?: number;
  location?: string;
  target_time_seconds?: number;
  notes?: string;
}

export interface RaceEventResponse extends RaceEventBase {
  id: number;
  user_id: number;
  created_at: string;
}

File: libs/shared/models/kotlin/Deportes.kt
package com.sentinel.shared.models

import java.time.Instant

/**
 * Shared Kotlin data classes for the Deportes (Sports) module.
 * Mirrors backend Pydantic schemas in apps/backend/src/models/schemas.py.
 */

// ---------------------------------------------------------------------------
// Sport
// ---------------------------------------------------------------------------

data class SportBase(
    val name: String,
    val icon: String? = null
)

data class SportCreate(
    val name: String,
    val icon: String? = null
)

data class SportResponse(
    val id: Int,
    val name: String,
    val icon: String? = null,
    val is_custom: Boolean,
    val created_at: Instant
)

// ---------------------------------------------------------------------------
// Sport Activity
// ---------------------------------------------------------------------------

data class SportActivityBase(
    val sport_id: Int,
    val date: Instant,
    val duration_seconds: Int,
    val distance_km: Double? = null,
    val calories: Int? = null,
    val heart_rate_avg: Int? = null,
    val extra_data: Map<String, Any>? = null,
    val notes: String? = null
)

data class SportActivityCreate(
    val sport_id: Int,
    val date: Instant,
    val duration_seconds: Int,
    val distance_km: Double? = null,
    val calories: Int? = null,
    val heart_rate_avg: Int? = null,
    val extra_data: Map<String, Any>? = null,
    val notes: String? = null
)

data class SportActivityUpdate(
    val date: Instant? = null,
    val duration_seconds: Int? = null,
    val distance_km: Double? = null,
    val calories: Int? = null,
    val heart_rate_avg: Int? = null,
    val extra_data: Map<String, Any>? = null,
    val notes: String? = null
)

data class SportActivityResponse(
    val id: Int,
    val user_id: Int,
    val sport_id: Int,
    val date: Instant,
    val duration_seconds: Int,
    val distance_km: Double? = null,
    val calories: Int? = null,
    val heart_rate_avg: Int? = null,
    val extra_data: Map<String, Any>? = null,
    val notes: String? = null,
    val created_at: Instant
)

// ---------------------------------------------------------------------------
// Sport Stats & Personal Records
// ---------------------------------------------------------------------------

data class SportStatsResponse(
    val sport_id: Int,
    val sport_name: String,
    val total_activities: Int,
    val total_time_seconds: Int,
    val total_distance_km: Double? = null,
    val average_duration_seconds: Double? = null
)

data class PersonalRecordResponse(
    val sport_id: Int,
    val best_time_seconds: Int? = null,
    val best_distance_km: Double? = null,
    val best_pace_seconds_per_km: Int? = null,
    val best_time_date: Instant? = null,
    val best_distance_date: Instant? = null
)

// ---------------------------------------------------------------------------
// Training Plan
// ---------------------------------------------------------------------------

data class TrainingPlanWeekUpdate(
    val week_number: Int,
    val target_sessions: Int,
    val completed_sessions: Int = 0,
    val notes: String? = null
)

data class TrainingPlanBase(
    val name: String,
    val description: String? = null,
    val start_date: Instant,
    val end_date: Instant
)

data class TrainingPlanCreate(
    val name: String,
    val description: String? = null,
    val start_date: Instant,
    val end_date: Instant,
    val weeks: List<TrainingPlanWeekUpdate>? = null
)

data class TrainingPlanUpdate(
    val name: String? = null,
    val description: String? = null,
    val start_date: Instant? = null,
    val end_date: Instant? = null,
    val weeks: List<TrainingPlanWeekUpdate>? = null
)

data class TrainingPlanWeekResponse(
    val id: Int,
    val week_number: Int,
    val target_sessions: Int,
    val completed_sessions: Int,
    val notes: String? = null
)

data class TrainingPlanResponse(
    val id: Int,
    val name: String,
    val description: String? = null,
    val start_date: Instant,
    val end_date: Instant,
    val created_at: Instant,
    val updated_at: Instant
)

data class TrainingPlanDetailResponse(
    val id: Int,
    val name: String,
    val description: String? = null,
    val start_date: Instant,
    val end_date: Instant,
    val created_at: Instant,
    val updated_at: Instant,
    val weeks: List<TrainingPlanWeekResponse> = emptyList()
)

// ---------------------------------------------------------------------------
// Race Event
// ---------------------------------------------------------------------------

data class RaceEventBase(
    val name: String,
    val sport_id: Int,
    val event_date: Instant,
    val distance_km: Double? = null,
    val location: String? = null,
    val target_time_seconds: Int? = null,
    val notes: String? = null
)

data class RaceEventCreate(
    val name: String,
    val sport_id: Int,
    val event_date: Instant,
    val distance_km: Double? = null,
    val location: String? = null,
    val target_time_seconds: Int? = null,
    val notes: String? = null
)

data class RaceEventUpdate(
    val name: String? = null,
    val sport_id: Int? = null,
    val event_date: Instant? = null,
    val distance_km: Double? = null,
    val location: String? = null,
    val target_time_seconds: Int? = null,
    val notes: String? = null
)

data class RaceEventResponse(
    val id: Int,
    val user_id: Int,
    val name: String,
    val sport_id: Int,
    val event_date: Instant,
    val distance_km: Double? = null,
    val location: String? = null,
    val target_time_seconds: Int? = null,
    val notes: String? = null,
    val created_at: Instant
)
